src/Shared/FileMatcher.cs

```diff
@@ -576,53 +620,58 @@ internal static bool IsDirectorySeparator(char c)
         /// </summary>
         /// <param name="paths">Paths to remove current directory from.</param>
         /// <param name="projectDirectory"></param>
-        internal static void RemoveProjectDirectory
+        internal static IEnumerable<string> RemoveProjectDirectory
         (
-            string[] paths,
+            IEnumerable<string> paths,
             string projectDirectory
         )
         {
             bool directoryLastCharIsSeparator = IsDirectorySeparator(projectDirectory[projectDirectory.Length - 1]);
-            for (int i = 0; i < paths.Length; i++)
+            foreach (string path in paths)
```

Comment:
This can throw.

Reply:
All of these will be caught via the aggregate exception in GetFiles right?

---
In reply to: [143076202](https://github.com/Microsoft/msbuild/pull/2572#discussion_r143076202) [](ancestors = 143076202)

Reply:
Actually, they will be, but this is not correct as we don't want to return "empty" result just because one of the subdirectories, failed to get its files. The catching of the `AggreateException` is supposed to do the same as the  [other catch](https://github.com/Microsoft/msbuild/blob/master/src/Shared/FileMatcher.cs#L1924), when the exception is thrown inside a `Parallel.ForEach`, which will always throw an `AggreateException` containing the actual exception. The iteration over `paths` should not throw for `SecurityException` and `UnauthorizedAccessException` in order to preserve the old behavior, I will correct this.